code/RealMeService.php

```diff
@@ -205,58 +233,189 @@ class RealMeService extends Object
     private static $metadata_contact_support_surname = null;
 
     /**
-     * @return bool true if the user is correctly authenticated, false if there was an error with login
+     * @var OneLogin_Saml2_Auth|null Set by {@link getAuth()}, which creates an instance of OneLogin_Saml2_Auth to check
+     * authentication against
+     */
+    private $auth = null;
+
+    /**
+     * @var string|null The last error message during login enforcement
+     */
+    private $lastError = null;
+
+    /**
+     * @return bool|null true if the user is correctly authenticated, false if there was an error with login
+     *
      * NB: If the user is not authenticated, they will be redirected to RealMe to login, so a boolean false return here
-     * indicates that there was a failure during the authentication process (perhaps a communication issue)
+     * indicates that there was a failure during the authentication process (perhaps a communication issue). You can
+     * check getLastError() to see if a human-readable error message exists for display.
      */
     public function enforceLogin()
     {
-        $auth = new SimpleSAML_Auth_Simple($this->config()->auth_source_name);
+        // First, check to see if we have an existing authenticated session
+        if($this->isAuthenticated()) {
+            return true;
+        }
+
+        // If not, attempt to retrieve authentication data from OneLogin (in case this is called during SAML assertion)
+        try {
+            $this->getAuth()->processResponse();
+            $errors = $this->getAuth()->getErrors();
+            $translatedMessage = null;
+
+            if(is_array($errors) && sizeof($errors) > 0) {
```

Comment:
`!empty()` is probably more efficient than counting `$errors` and evaluating against `>0`